src/dataflow-types/src/client/controller/storage.rs

```diff
@@ -666,3 +667,109 @@ impl<T: Timestamp> CollectionState<T> {
         }
     }
 }
+
+/// A wrapper that holds on to backing persist shards/collections that the
+/// storage controller is aware of. The handles hold back the since frontier and
+/// we need to downgrade them when the read capabilities change.
+///
+/// Internally, this has an async task and the methods for registering a handle
+/// and downgrading sinces add commands to a queue that this task is working
+/// off. This makes the methods non-blocking and moves the work outside the main
+/// coordinator task, meaning the coordinator is spending less time waiting on
+/// persist calls.
+#[derive(Debug)]
+struct PersistWorker<T: Timestamp + Lattice + Codec64> {
+    tx: UnboundedSender<PersistWorkerCmd<T>>,
+    handle: Option<JoinHandle<Result<(), anyhow::Error>>>,
+}
+
+impl<T> Drop for PersistWorker<T>
+where
+    T: Timestamp + Lattice + Codec64,
+{
+    fn drop(&mut self) {
+        self.send(PersistWorkerCmd::Shutdown);
+
+        let handle = match Handle::try_current() {
+            Ok(x) => x,
+            Err(_) => {
+                warn!("PersistDowngrade dropped without a tokio Runtime present, cannot join on handle");
+                return;
+            }
+        };
+
+        let res = handle
+            .block_on(self.handle.take().expect("missing join handle"))
+            .expect("join error");
+
+        match res {
+            Ok(()) => (), // All good!
+            Err(e) => {
+                error!("error while shutting down PersistDowngrade: {}", e)
+            }
+        }
+    }
+}
+
+/// Commands for [PersistWorker].
+#[derive(Debug)]
+enum PersistWorkerCmd<T: Timestamp + Lattice + Codec64> {
+    Register(ShardId, ReadHandle<SourceData, (), T, Diff>),
+    Downgrade(ShardId, Antichain<T>),
+    Shutdown,
+}
+
+impl<T: Timestamp + Lattice + Codec64> PersistWorker<T> {
+    fn new() -> Self {
+        let (tx, mut rx) = tokio::sync::mpsc::unbounded_channel();
+
+        let handle = mz_ore::task::spawn(|| "PersistDowngrade", async move {
+            let mut read_handles: HashMap<ShardId, ReadHandle<SourceData, (), T, Diff>> =
+                HashMap::new();
+
+            while let Some(cmd) = rx.recv().await {
+                match cmd {
+                    PersistWorkerCmd::Register(shard_id, read_handle) => {
+                        let previous = read_handles.insert(shard_id, read_handle);
+                        if previous.is_some() {
+                            panic!("already registered a ReadHandle for shard {:?}", shard_id);
```

Comment:
this feels a bit panic-happy to me. IMO at most we should panic with soft_assertions on and in just replace the old one